app/assets/javascripts/controllers/topology/container_topology_controller.js

```diff
@@ -13,49 +14,44 @@ function ContainerTopologyCtrl($scope, $http, $interval, topologyService, $windo
 
   topologyService.mixinContextMenu(vm, vm);
 
-  ManageIQ.angular.rxSubject.subscribe(function(event) {
-    if (event.name === 'refreshTopology') {
-      vm.refresh();
-    }
-  });
-
-  vm.refresh = function() {
-    var id;
-    var type;
-    var arr;
-    var pathname = $window.location.pathname.replace(/\/$/, '');
-    if (pathname.match('/container_topology/show$')) {
-      // specifically the container_topology page - all container ems's
-      type = 'container_topology';
-      id = '';
-    } else if (pathname.match('/(.+)/show/([0-9]+)')) {
-      // any container entity except the ems
-      // search for pattern ^/<controller>/show/<id>$ in the pathname - /container_project/show/11
-      arr = pathname.match('/(.+)/show/([0-9]+)');
-      type = arr[1] + '_topology';
-      id = '/' + arr[2];
-    } else if (pathname.match('/(.+)/([0-9]+)')) {
-      // single entity topology of ems_container
-      // search for pattern ^/<controller>/<id>$ in the pathname - /ems_container/4
-      arr = pathname.match('/(.+)/([0-9]+)');
-      type = 'container_topology';
-      id = '/' + arr[2];
-    }
-
-    var url = '/' + type + '/data' + id;
-
-    $http.get(url)
-      .then(getContainerTopologyData)
-      .catch(miqService.handleFailure);
-  };
-
   vm.checkboxModel = {
     value: false
   };
 
   vm.legendTooltip = __("Click here to show/hide entities of this type");
 
+  vm.getTopologyData = function(response) {
+    var data = response.data;
+
+    var currentSelectedKinds = vm.kinds;
+
+    vm.items = data.data.items;
+    vm.relations = data.data.relations;
+    // NOTE: $scope.kinds is required by kubernetes-topology-icon
+    vm.kinds = $scope.kinds = data.data.kinds;
+    icons = data.data.icons;
+
+    if (currentSelectedKinds && (Object.keys(currentSelectedKinds).length !== Object.keys(vm.kinds).length)) {
+      vm.kinds = $scope.kinds = currentSelectedKinds;
+    } else if (data.data.settings && data.data.settings.containers_max_items) {
```

Comment:
Up until this `else` bit, this is exactly `topologyService#mixinRefresh#getTopologyData` .. 

Any chance we could share that bit too?

Or, another option .. move the else part to the generic code as well, letting the controller provide the list of hierarchies (and the value or name of `container_max_items`) - because the idea that we want to hide some kinds of items if there are too many sounds quite valid :) (and could be useful for networks as well)

Reply:
the `getTopologyData` is actually inside the service. Problem is that one topology controllers (i think it is container_topology controller) had different implementation than the others. I didn't had time to check differences in those functions.